streamer/src/nonblocking/connection_rate_limiter.rs

```diff
@@ -72,10 +79,157 @@ impl TotalConnectionRateLimiter {
+    state: Mutex<TokenBucketState>,
+}
+
+impl Clone for TokenBucket {
+    fn clone(&self) -> Self {
+        Self {
+            tokens_per_second: self.tokens_per_second,
+            max_tokens: self.max_tokens,
+            base_time: self.base_time,
+            state: Mutex::new(self.state.lock().unwrap().clone()),
+        }
+    }
+}
+
+impl TokenBucket {
+    /// Allocate a new TokenBucket
+    pub fn new(initial_tokens: u64, max_tokens: u64, tokens_per_second: f64) -> Self {
+        assert!(
+            tokens_per_second > 0.0,
+            "Token bucket can not have zero influx rate"
+        );
+        assert!(
+            initial_tokens <= max_tokens,
+            "Can not have more initial tokens than max tokens"
+        );
+        let base_time = Instant::now();
+        TokenBucket {
+            tokens_per_second,
+            max_tokens,
+            state: Mutex::new(TokenBucketState {
+                tokens: initial_tokens,
+                last_access: 0,
+            }),
+            base_time,
+        }
+    }
+
+    pub fn current_tokens(&self) -> u64 {
+        let mut state = self.state.lock().unwrap();
+        let now = self.time_us();
+        self.update_state(now, &mut state);
+        state.tokens
+    }
+
+    /// Attempts to consume tokens from bucket.
+    /// On success, returns Ok(amount of tokens left in the bucket)
+    /// On failure, returns Err(amount of tokens missing to fill request)
+    pub fn consume_tokens(&self, request_size: u64) -> Result<u64, u64> {
+        let mut state = self.state.lock().unwrap();
+        let now = self.time_us();
+        self.update_state(now, &mut state);
+        if state.tokens >= request_size {
+            state.tokens -= request_size;
+            Ok(state.tokens)
+        } else {
+            Err(request_size - state.tokens)
+        }
+    }
+
+    /// Retrieves monotonic time since bucket
+    /// creation
+    fn time_us(&self) -> u64 {
+        let now = Instant::now();
+        let elapsed = now.saturating_duration_since(self.base_time);
+        elapsed.as_micros() as u64
+    }
+
+    /// Updates internal state of the bucket by
+    /// depositing new tokens (if appropriate)
+    fn update_state(&self, now: u64, state: &mut TokenBucketState) {
+        debug_assert!(now >= state.last_access);
+        let elapsed = (now - state.last_access) as f64;
+        let new_tokens = elapsed * self.tokens_per_second / 1e6;
+        // check if we can mint at least 1 new token
+        if new_tokens > 1.0 {
+            // update time of last mint
+            state.last_access = now;
+            // fill the bucket
+            state.tokens = self
+                .max_tokens
+                .min(state.tokens.saturating_add(new_tokens as u64));
+        }
+    }
+}
+
+/// Provides rate limiting for multiple source IP addresses
+/// on demand. Uses LazyLru under the hood.
+pub struct KeyedRateLimiter {
+    data: RwLock<LruCache<IpAddr, TokenBucket>>,
```

Comment:
yeah, governer is using `DashMap<IpAddr, Atomicu64>` and i think we do something similar here for the performance sake (depends on when do we need this, of course).